### PR TITLE
feat: Add `with_attributes` context propagation for PG instrumentation

### PR DESCRIPTION
**Pull request by yoheyk**
Adding the `with_attributes` feature to PG instrumentation (e.g. mysql2: https://github.com/open-telemetry/opentelemetry-ruby/pull/1175)

This is going to be the forth one (we already have this feature in `Redis`, `HTTP::ClientContext` and `mysql2`) so still under the threshold of duplication 😄 : https://github.com/open-telemetry/opentelemetry-ruby/pull/1157#issuecomment-1089070239

https://github.com/open-telemetry/opentelemetry-ruby/discussions/1341

### CODE REVIEW

**Review by ahayworth — COMMENTED** [review]
Do we need `context_with_attributes` ? I don't see it actually being used here, and it seems like maybe it was copied-and-pasted from redis -> mysql2 -> this PR.

I'm wondering if we should just drop it or not.

**Comment by fbogsany** [comment]
There's no strong reason for `context_with_attributes` other than consistency with the [`Trace` context helpers](https://github.com/open-telemetry/opentelemetry-ruby/blob/6fed21de005917f1900a8b8dc879f89623e4c61a/api/lib/opentelemetry/trace.rb#L46-L73):
```ruby
    # Returns the current span from the current or provided context
    #
    # @param [optional Context] context The context to lookup the current
    #   {Span} from. Defaults to Context.current
    def current_span(context = nil)
      context ||= Context.current
      context.value(CURRENT_SPAN_KEY) || Span::INVALID
    end


    # Returns a context containing the span, derived from the optional parent
    # context, or the current context if one was not provided.
    #
    # @param [optional Context] context The context to use as the parent for
    #   the returned context
    def context_with_span(span, parent_context: Context.current)
      parent_context.set_value(CURRENT_SPAN_KEY, span)
    end


    # Activates/deactivates the Span within the current Context, which makes the "current span"
    # available implicitly.
    #
    # On exit, the Span that was active before calling this method will be reactivated.
    #
    # @param [Span] span the span to activate
    # @yield [span, context] yields span and a context containing the span to the block.
    def with_span(span)
      Context.with_value(CURRENT_SPAN_KEY, span) { |c, s| yield s, c }
    end
```

In practice, none of the copies of `context_with_attributes` is actually used anywhere AFAICT. I don't think consistency is a reasonable argument for keeping it. I think we should remove it from this PR and 🤔 about a deprecation process for the other instances (pragmatically, we can rely on semver, but we want to be cautious about declaring any instrumentation to be 1.0).

**Comment by yoheyk** [comment]
Thanks!! I'm going to remove `context_with_attributes` from this PR

File: instrumentation/pg/lib/opentelemetry/instrumentation/pg/patches/connection.rb [hunk 1]
```diff
@@ -85,6 +85,7 @@ def span_attrs(kind, *args)
 
             attrs = { 'db.operation' => validated_operation(operation), 'db.postgresql.prepared_statement_name' => statement_name }
             attrs['db.statement'] = sql unless config[:db_statement] == :omit
+            attrs.merge!(OpenTelemetry::Instrumentation::PG.attributes)
```
[the remark this answers is not in the document]
Comment (yoheyk) [reply instrumentation/pg/lib/opentelemetry/instrumentation/pg/patches/connection.rb:88]: > check the last line of this method

in this line, `attrs` overwrites `client_attributes` so I think this code is working as expected: 

https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/278a0b8d70136412ee7d8b99f1be94699f5de6cd/instrumentation/pg/lib/opentelemetry/instrumentation/pg/patches/connection.rb#L91

I added some test cases: 9ca6a7f3303aaa8752d9f6b6a8c457f329d2e595

But let me know if I should rewrite to improve readability!

[the remark this answers is not in the document]
Comment (ahayworth) [reply instrumentation/pg/lib/opentelemetry/instrumentation/pg/patches/connection.rb:88]: Oh, you know what .. you're right. 🤦

**Review by ahayworth — APPROVED** [review]
Welp, I'm a dingus - you were absolutely right the first time. I think I got my merge-direction confused!

This LGTM, and thank you so much for the contribution! 😄 😄

**Review by plantfansam — APPROVED** [review]

**Pull request merged**